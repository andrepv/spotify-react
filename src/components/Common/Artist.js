import React, { Component } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import musician from "images/musician.png";
import { ARTIST } from "constants/RouteConstants";
import LazyLoad from "components/Common/LazyLoad";
import "styles/Artist.scss";

export default class Artist extends Component {
  shouldComponentUpdate(nextProps) {
    return this.props.id !== nextProps.id;
  }

  render() {
    const {image, id, handler, name} = this.props;
    return (
      <div className="artist">
        <LazyLoad className="artist__pic">
          <div
            className={`artist__pic artist__pic_root bg-${
              !image ? "empty" : "center"
            }`}
            style={{backgroundImage: `url(${image || musician})`}}
          ></div>
        </LazyLoad>
        <Link
          to={`${ARTIST}/${id}`}
          className="artist__name"
        >
          <span onClick={handler}>
            {name}
          </span>
        </Link>
      </div>
    );
  }
}

Artist.propTypes = {
  id: PropTypes.string.isRequired,
  image: PropTypes.string,
  name: PropTypes.string.isRequired,
  handler: PropTypes.func,
};