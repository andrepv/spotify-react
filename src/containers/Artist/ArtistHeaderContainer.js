import { connect } from "react-redux";
import {
  loadArtist,
  followArtist,
  unfollowArtist,
} from "actions/ArtistActions";

const mapStateToProps = store => {
  return {
    artist: store.artist,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    loadArtist: artistId => {
      dispatch(loadArtist(artistId));
    },
    followArtist: artist => {
      dispatch(followArtist(artist));
    },
    unfollowArtist: artistId => {
      dispatch(unfollowArtist(artistId));
    },
  };
};

export const connectArtistHeader = connect(
  mapStateToProps,
  mapDispatchToProps
);