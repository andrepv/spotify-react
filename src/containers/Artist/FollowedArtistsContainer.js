import { connect } from "react-redux";
import {
  loadFollowedArtists,
  loadMoreFollowedArtists,
} from "actions/ArtistActions";

const mapStateToProps = store => {
  return {
    artists: store.followedArtists,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    loadArtists: () => {
      dispatch(loadFollowedArtists());
    },
    loadMore: lastArtistId => {
      dispatch(loadMoreFollowedArtists(lastArtistId));
    },
  };
};

export const connectFollowedArtists = connect(
  mapStateToProps,
  mapDispatchToProps
);