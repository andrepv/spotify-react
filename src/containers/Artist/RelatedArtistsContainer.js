import { connect } from "react-redux";
import { loadRelatedArtists } from "actions/ArtistActions";

const mapStateToProps = store => {
  return {
    artists: store.relatedArtists,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    loadRelatedArtists: artistId => {
      dispatch(loadRelatedArtists(artistId));
    },
  };
};

export const connectRelatedArtists = connect(
  mapStateToProps,
  mapDispatchToProps
);