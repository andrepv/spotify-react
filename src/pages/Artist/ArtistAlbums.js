import React, { Component } from "react";
import PropTypes from "prop-types";

import Carousel from "components/Common/Carousel";
import { ALBUMS, ARTIST } from "constants/RouteConstants";
import { connectArtistAlbums } from "containers/Album/ArtistAlbumsContainer";

export class ArtistAlbums extends Component {
  componentDidMount() {
    this.props.loadAlbums(this.props.id);
  }

  componentDidUpdate(prevProps) {
    const {id, loadAlbums} = this.props;
    if (id !== prevProps.id) {
      loadAlbums(id);
    }
  }

  render() {
    const {pending, items, total} = this.props.albums;
    const blockHeader = {title: "Albums"};
    if (items.length < total) {
      blockHeader.link = `${ARTIST}/${this.props.id}${ALBUMS}`;
    }
    if (!total) {
      return null;
    }
    return (
      <Carousel
        pending={pending}
        items={items}
        type='album'
        blockHeader={blockHeader}
      />
    );
  }
}

ArtistAlbums.propTypes = {
  albums: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  loadAlbums: PropTypes.func.isRequired,
};

export default connectArtistAlbums(ArtistAlbums);