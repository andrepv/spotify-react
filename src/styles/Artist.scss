@import
"variables",
"modularscale";

.artist {
  &__pic {
    border-radius: 100%;
    box-shadow: $shadows;
    width: 100%;
    margin-bottom: 10px;
    padding-bottom: 100%;
  }
  &__name {
    font-weight: 500;
    text-align: center;
    width: 100%;
    display: inline-block;
    color: $text-dark;
  }
  &-header {
    min-height: 350px;
    position: relative;
    text-align: center;
    margin-bottom: 40px;
    &:before {
      content: '';
      position: absolute;
      background: #000;
      left: 0;
      right: 0;
      width: 100%;
      height: 100%;
      opacity: .5;
    }
    &__info {
      z-index: 2;
    }
    &__name {
      margin-bottom: 20px;
      color: #fff;
    }
    &__followers {
      font-size: ms(-1);
      color: #fff;
    }
    &__btn.btn {
      width: 160px;
      margin: 0 auto;
      color: #fff;
    }
  }
}

.artists {
  display: flex;
  flex-wrap: wrap;
  margin-right: -30px;
  margin-bottom: -30px;
  .artist {
    margin-bottom: 30px;
    width: calc(100% / 5 - 30px);
    margin-right: 30px;
    @media (max-width: 900px) {
      width: calc(100% / 4 - 30px);
    }
    @media (max-width: 700px) {
      width: calc(100% / 3 - 30px);
    }
    @media (max-width: 550px) {
      width: calc(100% / 2 - 30px);
    }
    &__name {
      font-size: ms(-1);
    }
  }
  &_cols-6 {
    .artist {
      width: calc(100% / 6 - 30px);
      @media (max-width: 900px) {
        width: calc(100% / 4 - 30px);
      }
      @media (max-width: 550px) {
        width: calc(100% / 3 - 30px);
      }
      @media (max-width: 450px) {
        width: calc(100% / 2 - 30px);
      }
    }
  }
}