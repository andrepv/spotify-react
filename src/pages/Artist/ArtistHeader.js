import React, { Component } from "react";
import PropTypes from "prop-types";
import { connectArtistHeader } from "containers/Artist/ArtistHeaderContainer";

export class ArtistHeader extends Component {
  componentDidMount() {
    this.props.loadArtist(this.props.id);
  }

  componentDidUpdate(prevProps) {
    const {id, loadArtist} = this.props;
    if (id !== prevProps.id) {
      loadArtist(id);
    }
  }

  follow = () => {
    this.props.followArtist(this.props.artist);
  }

  unfollow = () => {
    this.props.unfollowArtist(this.props.id);
  }

  render() {
    const {image, name, followers, isFollower} = this.props.artist;
    return (
      <header
        className="artist-header bg-center flex-center"
        style={{backgroundImage: `url(${image})`}}
      >
        <div className="artist-header__info">
          <p className="artist-header__followers">{followers}</p>
          <h1 className="artist-header__name">{name}</h1>
          <button
            className="artist-header__btn btn"
            onClick={isFollower ? this.unfollow : this.follow}
          >
            {isFollower ? "Unfollow" : "Follow"}
          </button>
        </div>
      </header>
    );
  }
}

ArtistHeader.propTypes = {
  id: PropTypes.string.isRequired,
  artist: PropTypes.object.isRequired,
  loadArtist: PropTypes.func.isRequired,
  followArtist: PropTypes.func.isRequired,
  unfollowArtist: PropTypes.func.isRequired,
};

export default connectArtistHeader(ArtistHeader);