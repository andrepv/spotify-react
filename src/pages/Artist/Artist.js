import React, { Component } from "react";
import { Route, Switch } from "react-router-dom";

import { ALBUMS, SINGLES, ARTIST } from "constants/RouteConstants";
import ArtistHeader from "./ArtistHeader";
import ArtistTopTracks from "./ArtistTopTracks";
import ArtistAlbums from "./ArtistAlbums";
import ArtistSingles from "./ArtistSingles";
import RelatedArtists from "./RelatedArtists";
import Albums from "../ArtistAlbums";
import Singles from "../ArtistSingles";
import "styles/Artist.scss";

export default class Artist extends Component {
  componentDidUpdate(prevProps) {
    if (!this.props.history.location.state) {
      window.scrollTo(0, 0);
    }
  }

  componentDidMount() {
    window.scrollTo(0, 0);
  }

  render() {
    const {id} = this.props.match.params;
    return (
      <Switch location={this.props.location}>
        <Route exact path={`${ARTIST}/${id}`}>
          <section className="section-separators">
            <ArtistHeader id={id} />
            <ArtistTopTracks id={id} />
            <ArtistAlbums id={id} />
            <ArtistSingles id={id} />
            <RelatedArtists id={id} />
          </section>
        </Route>
        <Route
          path={`${ARTIST}/${id}${ALBUMS}`}
          render={() => {
            return <Albums id={id} />;
          }}
        />
        <Route
          path={`${ARTIST}/${id}${SINGLES}`}
          render={() => {
            return <Singles id={id} />;
          }}
        />
      </Switch>
    );
  }
}