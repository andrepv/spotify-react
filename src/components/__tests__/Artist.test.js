import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import Artist from "components/Common/Artist";
import musician from "images/musician.png";

describe("<Artist />", () => {
  it("renders correctly", () => {
    const props = {
      id: "1",
      name: "test",
      handler: jest.fn(),
      image: "test.png",
    };
    const component = renderer.create(
      <Router><Artist {...props}/></Router>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("check the component without an image", () => {
    const props = {
      id: "1",
      name: "test",
      image: "",
    };
    const wrapper = shallow(<Artist {...props} />);
    const target = wrapper.find(".artist__pic_root");
    expect(target.hasClass("bg-empty")).toEqual(true);
    expect(target.prop("style")).toHaveProperty(
      "backgroundImage", `url(${musician})`
    );
  });
});