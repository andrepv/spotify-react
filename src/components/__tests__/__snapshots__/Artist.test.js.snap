// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Artist /> renders correctly 1`] = `
<div
  className="artist"
>
  <div
    className="artist__pic artist__pic_root bg-center"
    style={
      Object {
        "backgroundImage": "url(test.png)",
      }
    }
  />
  <a
    className="artist__name"
    href="/artist/1"
    onClick={[Function]}
  >
    <span
      onClick={[MockFunction]}
    >
      test
    </span>
  </a>
</div>
`;
